File: leetcode-algorithms/542. 01 Matrix/542.01-matrix.cpp
/*
 * @lc app=leetcode id=542 lang=cpp
 *
 * [542] 01 Matrix
 *
 * https://leetcode.com/problems/01-matrix/description/
 *
 * algorithms
 * Medium (36.50%)
 * Likes:    811
 * Dislikes: 84
 * Total Accepted:    53.2K
 * Total Submissions: 144.5K
 * Testcase Example:  '[[0,0,0],[0,1,0],[0,0,0]]'
 *
 * Given a matrix consists of 0 and 1, find the distance of the nearest 0 for
 * each cell.
 * 
 * The distance between two adjacent cells is 1.
 * 
 * 
 * 
 * Example 1: 
 * 
 * 
 * Input:
 * [[0,0,0],
 * ⁠[0,1,0],
 * ⁠[0,0,0]]
 * 
 * Output:
 * [[0,0,0],
 * [0,1,0],
 * [0,0,0]]
 * 
 * 
 * Example 2: 
 * 
 * 
 * Input:
 * [[0,0,0],
 * ⁠[0,1,0],
 * ⁠[1,1,1]]
 * 
 * Output:
 * [[0,0,0],
 * ⁠[0,1,0],
 * ⁠[1,2,1]]
 * 
 * 
 * 
 * 
 * Note:
 * 
 * 
 * The number of elements of the given matrix will not exceed 10,000.
 * There are at least one 0 in the given matrix.
 * The cells are adjacent in only four directions: up, down, left and right.
 * 
 * 
 */
class Solution
{
public:
    vector<vector<int>> updateMatrix(vector<vector<int>> &matrix)
    {
        if (matrix.size() == 0)
        {
            return matrix;
        }
        for (int i = 0; i < matrix.size(); i++)
        {
            for (int j = 0; j < matrix[0].size(); j++)
            {
                if (matrix[i][j] == 1 && !hasNeiberZero(i, j, matrix))
                {
                    matrix[i][j] = INT_MAX;
                }
            }
        }
        for (int i = 0; i < matrix.size(); i++)
        {
            for (int j = 0; j < matrix[0].size(); j++)
            {
                if (matrix[i][j] == 1)
                {
                    dfs(matrix, i, j, 1);
                }
            }
        }
        return matrix;
    }
    void dfs(vector<vector<int>> &matrix, int row, int col, int val)
    {
        if (row < 0 || col < 0 || col >= matrix[0].size() || row >= matrix.size() || (matrix[row][col] <= val && val != 1))
        {
            return;
        }
        matrix[row][col] = val;
        dfs(matrix, row + 1, col, matrix[row][col] + 1);
        dfs(matrix, row - 1, col, matrix[row][col] + 1);
        dfs(matrix, row, col + 1, matrix[row][col] + 1);
        dfs(matrix, row, col - 1, matrix[row][col] + 1);
    }

    bool hasNeiberZero(int row, int col, vector<vector<int>> &matrix)
    {
        if (row > 0 && matrix[row - 1][col] == 0)
        {
            return true;
        }
        if (row < matrix.size() - 1 && matrix[row + 1][col] == 0)
        {
            return true;
        }
        if (col > 0 && matrix[row][col - 1] == 0)
        {
            return true;
        }
        if (col < matrix[0].size() - 1 && matrix[row][col + 1] == 0)
        {
            return true;
        }
        return false;
    }
};

File: leetcode-algorithms/542. 01 Matrix/542.01-matrix.py
#
# @lc app=leetcode id=542 lang=python3
#
# [542] 01 Matrix
#
# https://leetcode.com/problems/01-matrix/description/
#
# algorithms
# Medium (37.24%)
# Likes:    894
# Dislikes: 90
# Total Accepted:    58.3K
# Total Submissions: 155.9K
# Testcase Example:  '[[0,0,0],[0,1,0],[0,0,0]]'
#
# Given a matrix consists of 0 and 1, find the distance of the nearest 0 for
# each cell.
# 
# The distance between two adjacent cells is 1.
# 
# 
# 
# Example 1: 
# 
# 
# Input:
# [[0,0,0],
# ⁠[0,1,0],
# ⁠[0,0,0]]
# 
# Output:
# [[0,0,0],
# [0,1,0],
# [0,0,0]]
# 
# 
# Example 2: 
# 
# 
# Input:
# [[0,0,0],
# ⁠[0,1,0],
# ⁠[1,1,1]]
# 
# Output:
# [[0,0,0],
# ⁠[0,1,0],
# ⁠[1,2,1]]
# 
# 
# 
# 
# Note:
# 
# 
# The number of elements of the given matrix will not exceed 10,000.
# There are at least one 0 in the given matrix.
# The cells are adjacent in only four directions: up, down, left and right.
# 
# 
#

# @lc code=start
class Solution:
    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:
        def hasNeiberZero(row, col, matrix):
            if row > 0 and matrix[row - 1][col] == 0:
                return True
            if (row < len(matrix) - 1 and matrix[row + 1][col] == 0):
                return True
            if (col > 0 and matrix[row][col - 1] == 0):
                return True
            if (col < len(matrix[0]) - 1 and matrix[row][col + 1] == 0):
                return True
            return False

        def dfs(matrix, row, col, val):
            if (row < 0 or col < 0 or col >= len(matrix[0]) or row >= len(matrix)
                    or (matrix[row][col] <= val and val != 1)):
                return
            matrix[row][col] = val

            dfs(matrix, row + 1, col, matrix[row][col] + 1)
            dfs(matrix, row - 1, col, matrix[row][col] + 1)
            dfs(matrix, row, col + 1, matrix[row][col] + 1)
            dfs(matrix, row, col - 1, matrix[row][col] + 1)

        if (len(matrix) == 0):
            return matrix
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if (matrix[i][j] == 1 and not hasNeiberZero(i, j, matrix)):
                    matrix[i][j] = 99999999
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if (matrix[i][j] == 1):
                    dfs(matrix, i, j, 1)
        return matrix
# @lc code=end

